### PR TITLE
feat: add a line item to an order edit

### PR DESCRIPTION
**Pull request by fPolic**
**What**
- Implement adding a line item to order (edit)

**How**
- _by implementing the following "flow"_
  - generate a line item
  - computing line item adjustments for that line item
  - creating tax lines
  - creating a change record

**Testing**
- **_integration tests_**
  - check if line item and order item change objects are created (with correct tax lines)
  - line item adjustments are generated if
    - fixed discount is applied to cart
    - percentage discount is applied
- **_unit tests_** 
  - ensure that methods from Inventory, LineItem, LineItemAdjustment etc. services are called 

---

RESOLVES CORE-495

### CODE REVIEW

File: packages/medusa/src/services/__tests__/order-edit.ts [hunk 1]
```diff
@@ -267,7 +288,9 @@ describe("OrderEditService", () => {
       let result
 
       beforeEach(async () => {
-        result = await orderEditService.requestConfirmation(orderEditId, {loggedInUser: userId})
+        result = await orderEditService.requestConfirmation(orderEditId, {
+          loggedInUser: userId,
```
Comment (adrien2p) [inline packages/medusa/src/services/__tests__/order-edit.ts:292]: todo: `loggedInUserId`

File: packages/medusa/src/services/__tests__/order-edit.ts [hunk 2]
```diff
@@ -298,7 +320,9 @@ describe("OrderEditService", () => {
...
```
Comment (adrien2p) [inline packages/medusa/src/services/__tests__/order-edit.ts:324]: todo: same here

File: packages/medusa/src/services/order-edit-item-change.ts [hunk 3]
```diff
@@ -68,6 +69,25 @@ export default class OrderEditItemChangeService extends TransactionBaseService {
     return itemChange
   }
 
+  async create(data: DeepPartial<OrderItemChange>): Promise<OrderItemChange> {
```
Comment (adrien2p) [inline packages/medusa/src/services/order-edit-item-change.ts:72]: suggestion: Not all the properties should be available, what about that (which is what I ve done in my pr):
```ts
export type CreateOrderEditItemChangeInput = {
  type: OrderEditItemChangeType
  order_edit_id: string
  original_line_item_id?: string
  line_item_id?: string
}
```

File: packages/medusa/src/services/order-edit.ts [hunk 4]
```diff
@@ -388,6 +411,85 @@ export default class OrderEditService extends TransactionBaseService {
...
+          "order.cart.shipping_methods",
+
+          "order.region",
+        ],
+      })
+
+      if (!OrderEditService.isOrderEditActive(orderEdit)) {
+        throw new MedusaError(
+          MedusaError.Types.NOT_ALLOWED,
+          `Can not add an item to the edit with status ${orderEdit.status}`
```
Comment (adrien2p) [inline packages/medusa/src/services/order-edit.ts:439]: suggestion: `Can not add an item on the order edit ${orderEditId} with the status ${orderEdit.status}`

File: packages/medusa/src/services/order-edit.ts [hunk 5]
```diff
@@ -388,6 +411,85 @@ export default class OrderEditService extends TransactionBaseService {
     return orderEdit
   }
 
+  async addLineItem(orderEditId: string, data: AddOrderEditLineItemInput) {
```
Comment (adrien2p) [inline packages/medusa/src/services/order-edit.ts:414]: suggestion: I think that once my pr will be merged, this method will be able to just call [that](https://github.com/medusajs/medusa/pull/2246/files#diff-a26e080c56da65f0067681a51f853510451acbbe4700216d78a333826a71357dR530) and then create the item change wdyt?

Comment (adrien2p) [inline packages/medusa/src/services/order-edit.ts:414]: todo: missing return type

File: packages/medusa/src/services/order-edit.ts [hunk 6]
```diff
@@ -388,6 +411,85 @@ export default class OrderEditService extends TransactionBaseService {
...
+
+      // 4. generate change record (with new line item)
+
+      await this.orderEditItemChangeService_.withTransaction(manager).create({
+        type: OrderEditItemChangeType.ITEM_ADD,
+        line_item_id: lineItem.id,
+        order_edit_id: orderEditId,
+      })
+
+      return this.retrieve(orderEditId)
```
Comment (adrien2p) [inline packages/medusa/src/services/order-edit.ts:489]: suggestion: I think that we can return void and not have to retrieve the order edit. the consumer can retrieve it later and often does. wdyt?

**Review by adrien2p — COMMENTED** [review]
Nice job 💪

File: packages/medusa/src/services/order-edit.ts [hunk 7]
```diff
@@ -388,6 +411,85 @@ export default class OrderEditService extends TransactionBaseService {
...
+
+      await this.lineItemAdjustmentService_
+        .withTransaction(manager)
+        .createAdjustments(orderEdit.order.cart, lineItem)
+
+      // 3. generate tax lines
+
+      const calcContext = await this.totalsService_
+        .withTransaction(manager)
+        .getCalculationContext(orderEdit.order)
```
Comment (adrien2p) [inline packages/medusa/src/services/order-edit.ts:475]: todo: the calculation context is missing the allocation on the newly created line item since it does not exists on the order yet. also the context could be computed only for the new line item

File: integration-tests/api/__tests__/admin/order-edit.js [hunk 8]
```diff
@@ -612,6 +615,233 @@ describe("[MEDUSA_FF_ORDER_EDITING] /admin/order-edits", () => {
...
+            }),
+          ]),
+          /*
+           * Computed totals are appended to the response
+           */
+          discount_total: 0,
+          gift_card_total: 0,
+          gift_card_tax_total: 0,
+          shipping_total: 0,
+          subtotal: 1000 + 2 * 200,
```
Comment (adrien2p) [inline integration-tests/api/__tests__/admin/order-edit.js:731]: question: How do you get the right amount since the tax line is associated to the new item but the item id is the original one, therefore I think that you got lucky because the allocation map contains the association of the tax line of the original item which is the same as the new one 😂 but in fact I am pretty sure it does not use the right tax lines on the new item.

Can you tell me what you think of that:
https://github.com/medusajs/medusa/pull/2246/files#diff-a26e080c56da65f0067681a51f853510451acbbe4700216d78a333826a71357dR584
https://github.com/medusajs/medusa/pull/2246/files#diff-a26e080c56da65f0067681a51f853510451acbbe4700216d78a333826a71357dR231 (this one is temporary until we agree on the way to do it)

Comment (adrien2p) [reply integration-tests/api/__tests__/admin/order-edit.js:731]: How did you get the right result since the tax lines are associated to the new line but in the items collection the item id is the one from the original one which is different from the one in the tax line. Therefore the totals service will not associate them during the computation.

File: packages/medusa/src/services/order-edit.ts (hunk 7 above)
Comment (fPolic) [reply packages/medusa/src/services/order-edit.ts:476]: yes good catch, I added your fix from the comment below for now 👍

File: integration-tests/api/__tests__/admin/order-edit.js [hunk 9]
```diff
@@ -612,6 +615,233 @@ describe("[MEDUSA_FF_ORDER_EDITING] /admin/order-edits", () => {
...
+          ],
+          /*
+           * Computed items are appended to the response
+           */
+          items: expect.arrayContaining([
+            // already existing item on the order
+            expect.objectContaining({
+              id: lineItemId1,
+              order_id: orderId1,
+              tax_lines: [],
```
[the remark this answers is not in the document]
Comment (fPolic) [reply integration-tests/api/__tests__/admin/order-edit.js:708]: because it was created with `simpleOrderFactory`

[the remark this answers is not in the document]
Comment (adrien2p) [reply integration-tests/api/__tests__/admin/order-edit.js:708]: Ahah sorry, use too pass a tax line 😂

File: integration-tests/api/__tests__/admin/order-edit.js [hunk 10]
```diff
@@ -612,6 +615,233 @@ describe("[MEDUSA_FF_ORDER_EDITING] /admin/order-edits", () => {
...
+            }),
+            // NEW line item
+            expect.objectContaining({
+              variant: expect.objectContaining({ id: toBeAddedVariantId }),
+              quantity: 2,
+              order_id: null, // <-- NOT associated with the order at this point
+              tax_lines: [
+                expect.objectContaining({
```
[the remark this answers is not in the document]
Comment (fPolic) [reply integration-tests/api/__tests__/admin/order-edit.js:716]: there is no "original" one since this is newly added item

[the remark this answers is not in the document]
Comment (adrien2p) [reply integration-tests/api/__tests__/admin/order-edit.js:716]: Yeah i ve got it just after and updated my other comment. 🤣i was to deep in the update just before

File: packages/medusa/src/api/routes/admin/order-edits/add-line-item.ts [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+ *   - lang: JavaScript
+ *     label: JS Client
+ *     source: |
+ *       import Medusa from "@medusajs/medusa-js"
+ *       const medusa = new Medusa({ baseUrl: MEDUSA_BACKEND_URL, maxRetries: 3 })
+ *       // must be previously logged in or use api token
+ *       medusa.admin.orderEdit.addLineItem({ variant_id, quantity })
+ *       .then(({ order_edit }) => {
+ *         console.log(order_edit.id);
+ *       });
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/order-edits/add-line-item.ts:27]: nit: remove `;`s

File: packages/medusa/src/api/routes/admin/order-edits/add-line-item.ts [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+import { EntityManager } from "typeorm"
+
+import { OrderEditService } from "../../../../services"
+import {
+  defaultOrderEditFields,
+  defaultOrderEditRelations,
+} from "../../../../types/order-edit"
+
+/**
+ * @oas [post] /order-edits/:id/items
```
Comment (srindom) [inline packages/medusa/src/api/routes/admin/order-edits/add-line-item.ts:12]:
```suggestion
 * @oas [post] /order-edits/{id}/items
```

and param should be documented

File: packages/medusa/src/services/order-edit.ts [hunk 13]
```diff
@@ -388,6 +411,86 @@ export default class OrderEditService extends TransactionBaseService {
...
 
+  async addLineItem(orderEditId: string, data: AddOrderEditLineItemInput) {
+    return await this.atomicPhase_(async (manager) => {
+      const lineItemServiceTx = this.lineItemService_.withTransaction(manager)
+
+      const orderEdit = await this.retrieve(orderEditId, {
+        relations: [
+          "order",
+
+          "order.cart",
```
Comment (srindom) [inline packages/medusa/src/services/order-edit.ts:422]: todo: avoid relying on the cart

File: packages/medusa/src/services/order-edit.ts [hunk 14]
```diff
@@ -388,6 +411,86 @@ export default class OrderEditService extends TransactionBaseService {
...
+        )
+      }
+
+      const regionId = orderEdit.order.region_id
+
+      // 0. check inventory
+
+      await this.inventoryService_
+        .withTransaction(manager)
+        .confirmInventory(data.variant_id, data.quantity)
```
Comment (srindom) [inline packages/medusa/src/services/order-edit.ts:449]: todo: remove this check. Since this is done by the store operator - who can view stock levels - we assume that they are aware of stock details and "give them full power".

We can always add this back in at a later point if relevant, but this is restricting things unnecessarily at this point.

File: packages/medusa/src/services/order-edit.ts [hunk 15]
```diff
@@ -388,6 +411,86 @@ export default class OrderEditService extends TransactionBaseService {
...
+      await this.lineItemAdjustmentService_
+        .withTransaction(manager)
+        .createAdjustments(orderEdit.order.cart, lineItem)
+
+      // 3. generate tax lines
+
+      orderEdit.order.items = [lineItem]
+      const calcContext = await this.totalsService_
+        .withTransaction(manager)
+        .getCalculationContext(orderEdit.order)
```
Comment (srindom) [inline packages/medusa/src/services/order-edit.ts:476]: Want to refer to the comment made here: https://github.com/medusajs/medusa/pull/2246#discussion_r978371683

I.e., make sure that adjustments are only created for percentage discounts

**Review by srindom — CHANGES_REQUESTED** [review]
Nice work! Maybe you and @adrien2p should figure out how to deal with the adjustments for fixed discounts, so you use the same approach for this PR and #2246 :)

File: packages/medusa/src/services/order-edit.ts [hunk 16]
```diff
@@ -328,6 +342,79 @@ export default class OrderEditService extends TransactionBaseService {
...
 
+  async addLineItem(
+    orderEditId: string,
+    data: AddOrderEditLineItemInput
+  ): Promise<void> {
+    return await this.atomicPhase_(async (manager) => {
+      const lineItemServiceTx = this.lineItemService_.withTransaction(manager)
+
+      const orderEdit = await this.retrieve(orderEditId, {
+        relations: ["order", "order.region"],
```
Comment (srindom) [inline packages/medusa/src/services/order-edit.ts:353]: suggestion: we can avoid the extra join here
```suggestion
        relations: ["order"],
```

File: packages/medusa/src/services/order-edit.ts [hunk 17]
```diff
@@ -328,6 +342,79 @@ export default class OrderEditService extends TransactionBaseService {
...
+      /**
+       * Create new line item and refresh adjustments for all cloned order edit items
+       */
+
+      const lineItemData = await lineItemServiceTx.generate(
+        data.variant_id,
+        regionId,
+        data.quantity,
+        {
+          metadata: data.metadata,
```
Comment (srindom) [inline packages/medusa/src/services/order-edit.ts:374]: todo: should pass `customer_id` so that price selection can work correctly

File: packages/medusa/src/services/order-edit.ts [hunk 18]
```diff
@@ -328,6 +342,79 @@ export default class OrderEditService extends TransactionBaseService {
...
+          order_edit_id: orderEditId,
+        }
+      )
+
+      let lineItem = await lineItemServiceTx.create(lineItemData)
+      lineItem = await lineItemServiceTx.retrieve(lineItem.id, {
+        relations: ["variant"],
```
Comment (srindom) [inline packages/medusa/src/services/order-edit.ts:381]: question: what is the point of joining the variant here?

**Review by srindom — CHANGES_REQUESTED** [review]

File: packages/medusa/src/services/order-edit.ts (hunk 18 above)
Comment (fPolic) [reply packages/medusa/src/services/order-edit.ts]: removed it since it's not needed

File: packages/medusa/src/types/order-edit.ts [hunk 19]
```diff
@@ -21,6 +28,7 @@ export const defaultOrderEditRelations: string[] = [
   "changes.line_item",
   "changes.original_line_item",
   "items",
+  "items.adjustments",
```
Comment (adrien2p) [inline packages/medusa/src/types/order-edit.ts:31]: question: should we return the adjustments here? I think that we never return them unless the client use the expand field

File: packages/medusa/src/services/__tests__/line-item.js [hunk 20]
```diff
@@ -407,6 +406,7 @@ describe("LineItemService", () => {
           metadata: {},
           should_merge: true,
           includes_tax: true,
+          order_edit_id: null,
```
Comment (adrien2p) [inline packages/medusa/src/services/__tests__/line-item.js:409]: todo: this field should not appear here as the flag is not enabled for the order edit feature in that describe.

File: packages/medusa/src/services/__tests__/line-item.js [hunk 21]
```diff
@@ -518,6 +518,7 @@ describe("LineItemService", () => {
           metadata: {},
           should_merge: true,
           includes_tax: false,
+          order_edit_id: null,
```
Comment (adrien2p) [inline packages/medusa/src/services/__tests__/line-item.js:521]: todo: same as the previous comment

File: packages/medusa/src/services/__tests__/line-item.js (hunk 20 above)
Comment (adrien2p) [reply packages/medusa/src/services/__tests__/line-item.js:409]: you can see that in the other pr we did not had to do that. what happened in your cases that you needed to add it?

File: packages/medusa/src/services/line-item-adjustment.ts [hunk 22]
```diff
@@ -163,7 +152,7 @@ class LineItemAdjustmentService extends BaseService {
    * @return the result of the delete operation
    */
   async delete(
-    selectorOrIds: string | string[] | FilterableLineItemAdjustmentProps
+    selectorOrIds: string | string[] | string[] | FilterableLineItemAdjustmentProps
```
Comment (adrien2p) [inline packages/medusa/src/services/line-item-adjustment.ts:155]: todo: remove the double `string[]`

File: packages/medusa/src/services/line-item.ts [hunk 23]
```diff
@@ -257,6 +258,7 @@ class LineItemService extends BaseService {
           is_giftcard: variant.product.is_giftcard,
           metadata: context?.metadata || {},
           should_merge: shouldMerge,
+          order_edit_id: context.order_edit_id || null,
```
Comment (adrien2p) [inline packages/medusa/src/services/line-item.ts:261]: todo: The order_edit_id should only be allowed under the flag

File: packages/medusa/src/services/order-edit.ts [hunk 24]
```diff
@@ -13,26 +14,31 @@ import {
...
   TaxProviderService,
   TotalsService,
 } from "./index"
-import { CreateOrderEditInput, UpdateOrderEditInput } from "../types/order-edit"
+import {
+  AddOrderEditLineItemInput,
+  CreateOrderEditInput,
+  UpdateOrderEditInput,
+} from "../types/order-edit"
 import region from "./region"
```
Comment (srindom) [inline packages/medusa/src/services/order-edit.ts:29]:
```suggestion
```

**Review by adrien2p — COMMENTED** [review]
Very nice 🎉  few comments

File: packages/medusa/src/types/order-edit.ts (hunk 19 above)
Comment (srindom) [reply packages/medusa/src/types/order-edit.ts:31]: These are fine to include :)

File: integration-tests/api/__tests__/admin/order-edit.js [hunk 25]
```diff
@@ -775,6 +775,327 @@ describe("[MEDUSA_FF_ORDER_EDITING] /admin/order-edits", () => {
...
+          shipping_total: 0,
+          subtotal: 500, // 1 * 100$ + 2 * 200$
+          discount_total: 100, // discount === 20%
+          tax_total: 40, // tax rate === 10%
+          total: 440,
+        })
+      )
+    })
+
+    it("adding line item to the order edit with will create adjustments for fixed discount case", async () => {
```
Comment (srindom) [inline integration-tests/api/__tests__/admin/order-edit.js:988]: praise: good job covering the key edge cases here! 🙏

**Review by srindom — APPROVED** [review]
LGTM - w. a small nit

File: packages/medusa/src/types/order-edit.ts (hunk 19 above)
Comment (adrien2p) [reply packages/medusa/src/types/order-edit.ts:31]: in that case, @fPolic could I ask you to update the integration tests related to the update line item to not get the order edit with the expand if that is alright, otherwise I ll do it in a separate pr :)

Comment (fPolic) [reply packages/medusa/src/types/order-edit.ts:31]: ofc, will do

File: integration-tests/api/__tests__/admin/order-edit.js [hunk 26]
```diff
@@ -2024,7 +2024,7 @@ describe("[MEDUSA_FF_ORDER_EDITING] /admin/order-edits", () => {
       )
 
       response = await api.get(
```
Comment (adrien2p) [inline integration-tests/api/__tests__/admin/order-edit.js:2026]: comment: If you could remove the get entirely as the update line item will return the order edit with the default relations which now includes the adjustments
